rust/core-lib/src/file.rs

```diff
@@ -186,7 +186,15 @@ where
 }
 
 fn try_save(path: &Path, text: &Rope, encoding: CharacterEncoding) -> io::Result<()> {
-    let mut f = File::create(path)?;
+    use std::fs;
+
+    let tmp_extension = path
+        .extension()
+        .and_then(|extension| extension.to_str())
+        .map_or_else(|| "swp".to_string(), |extension| format!("{}.swp", extension));
```

Comment:
You can use `extension.to_string() + ".swp"` which is faster than `format!()`. This also allows you to turn `extension` into a `String` earlier, which in turn simplifies this line a bit.

Reply:
Made the first change, but I couldn't figure out a good way to do the conversion to `String` earlier-- since `extension` is an `OsStr`, it's a bit of a pain.

Reply:
I think you can do something like
```rust
path
.extension()
.and_then(|extension| extension.to_str()?.to_string())
.map_or_else(|| "swp".to_string(), |extension| extension+".swp")
```

If the `to_str()` fails, `?` will just return `None` from the closure.

Reply:
Unfortunately, `to_string()` returns `String` instead of `Option<String>` so we can't return like that. However, we _could_ add an `into()` call to coerce it into a `Some`, which is pretty neat!